leasing/report/invoice/invoicing_review.py

```diff
@@ -362,6 +366,147 @@ def get_incorrect_management_shares_data(self, service_unit_ids, cursor):
 
         return data
 
+    # TODO: Optimise this function and remove noqa
+    def get_gaps_in_billing_periods_data(self, service_unit_ids, cursor):  # noqa: C901
+
+        unique_lease_ids_query = """
+            SELECT li.identifier AS "lease_identifier",
+                l.id AS "lease_id",
+                l.start_date,
+                l.end_date
+            FROM leasing_lease l
+            INNER JOIN leasing_leaseidentifier li
+                ON l.identifier_id = li.id
+            INNER JOIN leasing_rent r
+                ON l.id = r.lease_id
+            INNER JOIN leasing_invoice i
+                ON l.id = i.lease_id
+            WHERE (l.end_date IS NULL OR l.end_date >= %(today)s)
+                AND (l.start_date IS NOT NULL AND l.start_date <= %(today)s)
+                AND l.service_unit_id = ANY(%(service_units)s)
+                AND l.deleted IS NULL
+                AND l.state NOT IN ('reservation', 'power_of_attorney')
+                AND r.deleted IS NULL
+                AND (r.start_date IS NOT NULL AND r.start_date <= %(today)s)
+                AND (r.end_date IS NULL OR r.end_date >= %(today)s)
+                AND i.deleted IS NULL
+                AND i.deleted_by_cascade IS false
+            ORDER BY li.identifier ASC;
+        """
+
+        rent_query = """
+            SELECT r.id AS "rent_id",
+                r.start_date AS "start_date",
+                r.end_date AS "end_date",
+                r.lease_id AS "lease_id"
+            FROM leasing_rent r
+            WHERE r.lease_id = ANY(%(unique_lease_ids)s)
+            ORDER BY r.lease_id ASC, r.created_at ASC;
+        """
+
+        invoice_query = """
+            SELECT i.id AS "invoice_id",
+                i.billing_period_start_date,
+                i.billing_period_end_date,
+                i.lease_id AS "lease_id"
+            FROM leasing_invoice i
+            WHERE i.lease_id = ANY(%(unique_lease_ids)s)
+            ORDER BY i.lease_id ASC, i.created_at ASC;
+        """
+
+        today = datetime.date.today()
+
+        cursor.execute(
+            unique_lease_ids_query, {"service_units": service_unit_ids, "today": today}
+        )
+        unique_lease_id_rows = dictfetchall(cursor)
+
+        unique_lease_ids = [row["lease_id"] for row in unique_lease_id_rows]
+
+        cursor.execute(rent_query, {"unique_lease_ids": unique_lease_ids})
+        rent_rows = dictfetchall(cursor)
+
+        cursor.execute(invoice_query, {"unique_lease_ids": unique_lease_ids})
+        invoice_rows = dictfetchall(cursor)
+
+        data = []
+
+        def get_active_rent_period_start_date(lease, first_rent):
```

Comment:
This is super confusing that there are functions inside another function between the functions code

Reply:
I wonder if the purpose is to use `today` as a global variable across the execution, and/or encapsulating the function signatures inside this file. These two inner functions could be moved outside of the surrounding function by:

- adding a third parameter for the datetime
- renaming them with 1 or 2 leading underscores to imply protected/private access modifier.

Reply:
... or just inline them, they are understandable oneliners anyway.